### PR TITLE
Refactor and test auto-upgrade functionality

### PR DESCRIPTION
**Pull request by thedadams**
These changes are broken up into ~3~ 4 commits for reviewability purposes. These changes should be squashed before merging.

The first commit is where most of the changes occur. There were three ways to kick-off an `sync` previously:
1. A ticker based on the default auto upgrade config value
2. An on-demand sync call
3. After the sync was over, a goroutine would wait until the nearest calculated next time to upgrade and call sync again. This step also took into account the default auto upgrade config value.

Since 1 and 3 were doing similar things, they were combined to simplify the package-level state and make the code more testable.

Additionally, if an auto-upgrade-enabed app was deleted and recreated before the auto-upgrade sync ran, the next upgrade time for that app would still exist for the old version of the app and a user would have to wait for the next interval before the app would completely deploy. Now we detect when an app is new and sync it immediately.

The second commit is ONLY separating the `sync` method into multiple functions. The code is copied and pasted and none of it is changed with the possible exception of variable names.

The third commit adds tests to the code. Some changes were also made to the implementation for testability and mockability purposes.

A fourth commit has been added to address acorn args validation. The ImageDetails api-server endpoint now supports auto-upgrade tags and will get the latest tag for the pattern and return the ImageDetails for that image.

Issues:
https://github.com/acorn-io/acorn/issues/1032
https://github.com/acorn-io/acorn/issues/1056

### CODE REVIEW

File: pkg/autoupgrade/daemon.go
Comment (thedadams) [inline pkg/autoupgrade/daemon.go:191]: Should `"acorn"` here be `imageKey.namespace` instead?

**Comment by thedadams** [comment]
Note to reviewers: I will be making a significant change to the last commit in this PR, as well as adding an additional commit. I will comment again when this is done.

File: pkg/autoupgrade/daemon_test.go
Comment (tylerslaton) [inline pkg/autoupgrade/daemon_test.go:21]: Just a thought, but maybe we could move this to the `testdata` package and rename this to be `mockDaemonClient`?

File: pkg/autoupgrade/daemon.go
Comment (tylerslaton) [inline pkg/autoupgrade/daemon.go:53]: Since this doesn't return an error any more, won't we start the controller with this sync never being started?

**Review by tylerslaton — COMMENTED** [review]
Looking really good. Did a first pass at this but will come back to more sometime today or tomorrow.

File: pkg/controller/config/autoupgrade.go
Comment (thedadams) [reply pkg/controller/config/autoupgrade.go:9]: They are required because this is passed as a `router.Handler`, but we can ignore them. I will update.

File: pkg/autoupgrade/daemon_test.go
Comment (thedadams) [reply pkg/autoupgrade/daemon_test.go:21]: There isn't a `testdata` package here, and this mock is only valuable for unit tests of the `daemonClient`.

I did rename the type to `mockDaemonClient` as you suggested.

File: pkg/autoupgrade/daemon.go
Comment (thedadams) [reply pkg/autoupgrade/daemon.go:53]: The timer will always be started because we will use the `config.DefaultImageCheckIntervalDefault` (which is `5m`) if there is an error getting the config or parsing the check interval from the config.

I did change this slightly because your comment made me realize that we shouldn't be returning `0` if we can get the config object. We should return the `config.DefaultImageCheckIntervalDefault` (`5m`). I updated as such.

File: pkg/autoupgrade/daemon_test.go
Comment (tylerslaton) [reply pkg/autoupgrade/daemon_test.go]: Yep, totally valid conclusion here.

File: pkg/autoupgrade/daemon.go
Comment (tylerslaton) [reply pkg/autoupgrade/daemon.go:53]: Ah right on. Thanks for clearing that up for me.

File: pkg/autoupgrade/daemon.go
Comment (cjellick) [inline pkg/autoupgrade/daemon.go:45]:
```suggestion
// StartSync starts the daemon. It watches for new sync events coming and ensures a sync is triggered
```

File: pkg/autoupgrade/tags.go
Comment (cjellick) [inline pkg/autoupgrade/tags.go:49]: why bother with both a public and private version? seems like the public is purely pass through, so why not _just_ have the public?

Comment (thedadams) [reply pkg/autoupgrade/tags.go:49]: I am using the `daemonClient` interface so I can mock it for unit tests. By having this exported version take a `kclient.Client`, I can keep the `daemonClient` an unexported implementation detail.

Comment (cjellick) [reply pkg/autoupgrade/tags.go:49]: ah, got it!

**Review by cjellick — CHANGES_REQUESTED** [review]
Looks great! Thanks for cleaning this code up.

I noticed a random pre-existing typo in a comment. Want to suck that change in?

File: pkg/autoupgrade/daemon.go
Comment (thedadams) [reply pkg/autoupgrade/daemon.go]: Updated.

**Comment by thedadams** [comment]
I updated the comment and squashed all the commits.

File: pkg/autoupgrade/daemon.go
Comment (thedadams) [reply pkg/autoupgrade/daemon.go]: @cjellick Before we merge, I just want to verify whether this is correct. With all the "project" work, maybe this should be `imageKey.namespace` (which is the `app.Namespace`)?

Comment (thedadams) [reply pkg/autoupgrade/daemon.go]: EDIT: This isn't relevant with the recent changes, and, based on my investigation, this should be `imageKey.namespace` in the "project" paradigm.

**Review by tylerslaton — APPROVED** [review]
Gave this a test locally and thing are looking fine there. Great work here, code looks good 👍🏻

**Review by cjellick — APPROVED** [review]

**Comment by cjellick** [comment]
looks good. will leave it to you to merge and assign over

**Pull request merged**
